### PR TITLE
 odo delete --all only deletes the config directory if its empty

### PR DESCRIPTION
**Pull request by girishramnani**
## What is the purpose of this change? What does it change?
`odo delete --all` deleted the config file and index file. Then if the config directory is empty only then its deleted.

## Was the change discussed in an issue?
fixes https://github.com/openshift/odo/issues/2109

## How to test changes?
- the below commands shouldn't delete the .odo dir
 ```
odo create nodejs
echo "hello" > .odo/temp.txt
odo delete --all 
```
- the below commands should delete the .odo dir
```
odo create nodejs
odo delete --all 
```

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -102,6 +123,7 @@ type proxyLocalConfig struct {
 // serialize it.
 type LocalConfigInfo struct {
 	Filename         string `yaml:"FileName,omitempty"`
+	fs               fs
```
Comment (girishramnani) [inline pkg/config/config.go]: I am usually against having attributes which allow us to test being present in the main code. But the alternative is to have a global `fs` variable which I feel is even worse.

**Comment by girishramnani** [comment]
v4.3 Integration test failing due to `odo version` output difference, will open a ticket
and benchmark failing due to 
```
Pushing to component javacomponent of type local
[odo]  •  Checking files for pushing  ...
[odo]  ✓  Checking files for pushing [705777ns]
[odo]  •  Waiting for component to start  ...
[odo]  ✗  Waiting for component to start [4m]
[odo]  ✗  waited 4m0s but couldn't find running pod matching selector: 'deploymentconfig=javacomponent-myapp'
```

**Comment by girishramnani** [comment]
added unit tests.

**Comment by girishramnani** [comment]
```
info: Loading sha256:567b3f3af1ec931277374dbd62671688bac1869a33c441d3350b87d0d56e876c service-ca-operator
error: unable to extract layer sha256:92232441b340e615540ac5687834210ec33c9546f868f6d3a3a38d90f2494311 from registry.svc.ci.openshift.org/ci-op-kmcjzmqh/stable@sha256:3e30cb5eda859b89c074341ae13d64da4296a8507b9ee55bc6849d4974756b4a: stream error: stream ID 137; INTERNAL_ERROR
```
/retest

File: pkg/config/config.go (hunk 1 above)
Comment (kadel) [reply pkg/config/config.go]: > I am usually against having attributes which allow us to test being present in the main code.

Agreed. This is not a good approach :-(
There has to be a better way to do this.
/hold

Comment (girishramnani) [reply pkg/config/config.go]: I can keep it global

Comment (girishramnani) [reply pkg/config/config.go]: One good thing about this approach is that default behavior is preserved

Comment (kadel) [reply pkg/config/config.go]: > I can keep it global

I'm afraid that that would make it even worse :-(

File: pkg/config/config.go [hunk 2]
```diff
@@ -259,9 +280,37 @@ func (lci *LocalConfigInfo) SetConfiguration(parameter string, value interface{}
...
+	_, err := lci.fs.Stat(configDir)
+	if os.IsNotExist(err) {
+		// if the config dir doesn't exist then we dont mind
+		return nil
+	} else if err != nil {
+		// possible to not have permission to the dir
+		return err
+	}
+
+	f, err := lci.fs.Open(configDir)
```
Comment (cdrage) [inline pkg/config/config.go]: Describe these lines of code :)

File: pkg/config/config.go [hunk 3]
```diff
@@ -259,9 +280,37 @@ func (lci *LocalConfigInfo) SetConfiguration(parameter string, value interface{}
...
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	_, err = f.Readdir(1)
```
Comment (cdrage) [inline pkg/config/config.go]: why (1) ?

File: pkg/config/config.go [hunk 4]
```diff
@@ -259,9 +280,37 @@ func (lci *LocalConfigInfo) SetConfiguration(parameter string, value interface{}
...
+	// if directory is empty we can remove it
+	if err == io.EOF {
```
Comment (cdrage) [inline pkg/config/config.go]: what about err != nil ?

Comment (cdrage) [reply pkg/config/config.go]: (anything else that's not io.EOF)

File: pkg/config/config.go [hunk 5]
```diff
@@ -102,6 +121,7 @@ type proxyLocalConfig struct {
 // serialize it.
 type LocalConfigInfo struct {
 	Filename         string `yaml:"FileName,omitempty"`
+	fs               fs
```
[the remark this answers is not in the document]
Comment (girishramnani) [reply pkg/config/config.go]: fs = filesystem isn't that a universally acceptable short form?

File: pkg/config/config.go [hunk 6]
```diff
@@ -25,13 +26,31 @@ const (
...
 )
 
+type fs interface {
+	Open(name string) (file, error)
+	Stat(name string) (os.FileInfo, error)
+	Remove(name string) error
+}
+
+// we need this file interface to mock the file system
+type file interface {
```
[the remark this answers is not in the document]
Comment (girishramnani) [reply pkg/config/config.go]: It matches the hidden `file`  interface but I can call it mockFile if you like

File: pkg/config/config.go (hunk 3 above)
Comment (girishramnani) [reply pkg/config/config.go]: Reads only at most 1 files from the dir to check id there is any

**Comment by girishramnani** [comment]
after some discussion with @kadel. We finally came to the conclusion that using a memory mapped filesystem would be a better approach

File: pkg/config/config.go [hunk 7]
```diff
@@ -259,9 +280,37 @@ func (lci *LocalConfigInfo) SetConfiguration(parameter string, value interface{}
 
 }
 
-// DeleteConfigDir Deletes the config directory with the config file
-func (lci *LocalConfigInfo) DeleteConfigDir() error {
-	return os.RemoveAll(filepath.Dir(lci.Filename))
+// DeleteConfigDirIfEmpty Deletes the config directory if its empty
+func (lci *LocalConfigInfo) DeleteConfigDirIfEmpty() error {
```
[the remark this answers is not in the document]
Comment (girishramnani) [reply pkg/config/config.go]: yes

File: pkg/config/config.go (hunk 4 above)
Comment (girishramnani) [reply pkg/config/config.go]: resolved

File: pkg/config/config.go (hunk 1 above)
Comment (girishramnani) [reply pkg/config/config.go]: /hold cancel
Now we are using afero with the best approach.

**Comment by girishramnani** [comment]
@cdrage resolved the comments and @amitkrout implemented unit tests

**Comment by cdrage** [comment]
The folder still doesn't delete if there's something in there.

Steps to replicate:

```sh
git clone https://github.com/openshift/nodejs-ex
cd nodejs-ex
odo create nodejs
odo push
touch foobar.txt
odo push
odo delete --all # does not delete the directory?
```


See:

```sh
~/nodejs-ex  master ✗                                                                                                                                                                                                                                                   398d ◒  ⍉
▶ odo delete --all -f
 ✗  Component nodejs-nodejs-ex-azwj does not exist on the cluster
 ✓  Config for the Component nodejs-nodejs-ex-azwj has been deleted
```

```sh
~/nodejs-ex  master ✗                                                                                                                                                                                                                                                    398d ◒  
▶ ls -lah
total 80K                                                                                                                                                                                                                                                                        
drwxr-xr-x   8 wikus wikus 4.0K Nov 15 16:42 .
drwxr-xr-x 404 wikus wikus  20K Nov 18 10:13 ..
drwxr-xr-x   8 wikus wikus 4.0K Nov 18 10:13 .git                                                                                                                                                                                                                                
-rw-r--r--   1 wikus wikus   28 Oct 28 10:50 .gitignore
drwxr-xr-x   3 wikus wikus 4.0K Oct 28 10:50 helm                
drwxr-x---   2 wikus wikus 4.0K Nov 18 10:13 .odo                                
drwxr-xr-x   4 wikus wikus 4.0K Oct 28 10:50 openshift
-rw-r--r--   1 wikus wikus  817 Oct 28 10:50 package.json                                                                                                                                                                                                                        
-rw-r--r--   1 wikus wikus  13K Oct 28 10:50 README.md
-rw-r--r--   1 wikus wikus 3.8K Oct 28 10:50 server.js           
drwxr-xr-x   2 wikus wikus 4.0K Oct 28 10:50 tests                 
drwxr-xr-x   2 wikus wikus 4.0K Oct 28 10:50 views
                                                                                                                                                                                                                                                                                 
~/nodejs-ex  master ✗                                                                                                                                                                                                                                                    398d ◒  
▶ cat .odo/odo-file-index.json                                   

~/nodejs-ex  master ✗                                                                                                                                                                                                                                                   398d ◒  ⍉
▶        
         
~/nodejs-ex  master ✗                                                                                                                                                                                                                                                   398d ◒  ⍉
▶ ls -lah .odo                                 
total 12K                                        
drwxr-x--- 2 wikus wikus 4.0K Nov 18 10:13 .           
drwxr-xr-x 8 wikus wikus 4.0K Nov 15 16:42 ..    
-rw-r--r-- 1 wikus wikus 2.6K Nov 15 16:42 odo-file-index.json
```

**Comment by girishramnani** [comment]
@cdrage added the source to the original code

**Comment by kadel** [comment]
Shouldn't `odo delete --all` also delete the file index?
/hold

we need integration tests to cover this

```
▶ ls .odo
config.yaml         odo-file-index.json

▶ odo delete --all
 ✗  Component nodejs-nodejs-ex-lerx does not exist on the cluster
? Are you sure you want to delete local config for nodejs-nodejs-ex-lerx? Yes
 ✓  Config for the Component nodejs-nodejs-ex-lerx has been deleted

▶ ls .odo
odo-file-index.json
```

**Comment by girishramnani** [comment]
> Shouldn't `odo delete --all` also delete the file index?
> /hold
> 
> we need integration tests to cover this
> 
> ```
> ▶ ls .odo
> config.yaml         odo-file-index.json
> 
> ▶ odo delete --all
>  ✗  Component nodejs-nodejs-ex-lerx does not exist on the cluster
> ? Are you sure you want to delete local config for nodejs-nodejs-ex-lerx? Yes
>  ✓  Config for the Component nodejs-nodejs-ex-lerx has been deleted
> 
> ▶ ls .odo
> odo-file-index.json
> ```

So there was a bug when the `componentContext` was going empty when no value was provided, due to this `indexfile`'s location wasn't getting resolved.
I added the necessary fix and integration test @kadel .

File: pkg/config/config_test.go [hunk 8]
```diff
@@ -439,3 +442,85 @@ func TestGetOSSourcePath(t *testing.T) {
 		})
 	}
 }
+
+func TestDeleteConfigDirIfEmpty(t *testing.T) {
+
+	t.Run("empty config dir", func(t *testing.T) {
```
Comment (kadel) [inline pkg/config/config_test.go]: Would it be possible to convert this those test to use a similar system of a table drive tests as in all other tests?

Comment (girishramnani) [reply pkg/config/config_test.go]: No actually, I tried that. The check and setup is different for both approach

**Comment by girishramnani** [comment]
```
2019/11/21 05:04:38 No custom metadata found and prow metadata already exists. Not updating the metadata.
2019/11/21 05:04:39 Ran for 2m58s
2019/11/21 05:04:40 Submitted failure event to sentry (id=0af94b82e9014debb55adcb79a15b41d)
error: could not run steps: step unit failed: failed to create or restart test pod: unable to create pod: timed out waiting for the condition
```

**Comment by girishramnani** [comment]
> Code & functionality LGTM!
> /lgtm

I haven't received any approvals @dharmit & @cdrage . 1st person who reviews the PR ( if is an approver ) then please approve instead of `lgtm`

**Comment by girishramnani** [comment]
```
level=error msg="Error: Error waiting for instance (i-06ba3dc8bf28f7dca) to become ready: Failed to reach target state. Reason: Server.InternalError: Internal error on launch"
level=error
level=error msg="  on ../tmp/openshift-install-551414786/master/main.tf line 89, in resource \"aws_instance\" \"master\":"
level=error msg="  89: resource \"aws_instance\" \"master\" {"
```
/retest

**Comment by dharmit** [comment]
> I haven't received any approvals @dharmit & @cdrage . 1st person who reviews the PR ( if is an approver ) approves. So please approve instead of `lgtm`

I'm not sure what `1st person who reviews the PR ( if is an approver ) approves` means (particularly "if is an approver") but I didn't approve because I found things OK and it's @kadel who had some comments on the PR. :smile:

**Comment by girishramnani** [comment]
understood, no problem :slightly_smiling_face:

**Pull request merged**